diff --git a/Makefile.in b/Makefile.in
index 309623be..f09c31d1 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -51,12 +51,13 @@ DIST_VERSION := $(DST_BASE)/version
 DEPLOY_DIR := ./deploy
 REDIST_DIR := ./redist
 
-i386_CFLAGS    := -mstackrealign -march=nocona -mtune=core-avx2 -mfpmath=sse
-x86_64_CFLAGS  := -mcmodel=small -march=nocona -mtune=core-avx2 -mfpmath=sse
+HOST_CFLAGS    ?= -O2 -march=nocona -mtune=core-avx2
+i386_CFLAGS    := -mstackrealign $(HOST_CFLAGS) -mfpmath=sse
+x86_64_CFLAGS  := -mcmodel=small $(HOST_CFLAGS) -mfpmath=sse
 aarch64_CFLAGS := -march=armv8.2-a -mtune=cortex-x3
 arm64ec_CFLAGS := -march=armv8.2-a -mtune=cortex-x3
 
-CFLAGS  = -O2 -fwrapv -fno-strict-aliasing
+CFLAGS  = -fwrapv -fno-strict-aliasing
 CFLAGS += -ggdb -ffunction-sections -fdata-sections -fno-omit-frame-pointer
 CFLAGS += -ffile-prefix-map=$(CCACHE_BASEDIR)=.
 
diff --git a/configure.sh b/configure.sh
index 60b6df43..3b5ca931 100755
--- a/configure.sh
+++ b/configure.sh
@@ -188,6 +188,8 @@ function configure() {
       echo "ENABLE_CCACHE := 1"
     fi
 
+    echo "HOST_CFLAGS := ${CFLAGS:--O2 -march=nocona -mtune=core-avx2}"
+
     # Include base
     echo ""
     echo "include \$(SRCDIR)/Makefile.in"
