diff --git a/Makefile.in b/Makefile.in
index 309623be..c5654d79 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -1409,6 +1409,30 @@ $(DIST_AV1_PATENTS): $(AV1_PATENTS)
 all-dist: $(DIST_LICENSE) $(DIST_OFL_LICENSE) $(DIST_AV1_PATENTS)
 
 
+##
+## protonfixes
+##
+
+$(eval $(call rules-source,protonfixes,$(SRCDIR)/protonfixes))
+$(eval $(call rules-common,protonfixes,x86_64,unix))
+
+$(OBJ)/.protonfixes-x86_64-build:
+	cd $(PROTONFIXES_SRC) && \
+	env $(PROTONFIXES_x86_64_ENV) BUILD=$(PROTONFIXES_x86_64_OBJ) TARGET_ARCH=$(TARGET_ARCH) make
+	cd $(PROTONFIXES_SRC) && \
+	env $(PROTONFIXES_x86_64_ENV) BUILD=$(PROTONFIXES_x86_64_OBJ) DIST=$(PROTONFIXES_x86_64_DST) TARGET_ARCH=$(TARGET_ARCH) make install
+	touch $@
+
+PROTONFIXES_TARGET := $(addprefix $(DST_BASE)/,protonfixes)
+
+$(OBJ)/.protonfixes-x86_64-dist:
+	mkdir -p $(PROTONFIXES_TARGET)
+	rsync --delete -arx $(PROTONFIXES_x86_64_DST)/ $(PROTONFIXES_TARGET)
+	touch $@
+
+all-dist: protonfixes
+
+
 ##
 ## proton(.py), filelock.py, etc.
 ##
diff --git a/proton b/proton
index 076bbe32..65b453cb 100755
--- a/proton
+++ b/proton
@@ -38,6 +38,8 @@ from ctypes import c_ssize_t
 from filelock import FileLock
 from random import randrange
 
+import protonfixes
+
 #To enable debug logging, copy "user_settings.sample.py" to "user_settings.py"
 #and edit it if needed.
 
@@ -2150,6 +2152,13 @@ if __name__ == "__main__":
 
     g_session.init_session(sys.argv[1] != "runinprefix")
 
+    protonfixes.setup(g_session.env, "PATH", ld_path_var, prepend_to_env_str)
+    protonfixes.setup_mount_drives(setup_dir_drive)
+    protonfixes.winetricks(g_session.env, g_proton.wine_bin, g_proton.wineserver_bin)
+    g_session.check_environment("PROTON_USE_OPTISCALER", "optiscaler")
+    protonfixes.setup_upscalers(g_session.compat_config, g_session.env, g_compatdata.base_dir, g_compatdata.prefix_dir)
+    protonfixes.execute()
+
     #determine mode
     rc = 0
     if sys.argv[1] == "run":
